### PR TITLE
Rules Base Sampler

### PR DESCRIPTION
**Pull request by martin308**
Adds new sampler type, which implements some basic rules based sampling.

### CODE REVIEW

File: config/file_config.go [hunk 1]
```diff
@@ -306,6 +322,8 @@ func (f *fileConfig) GetSamplerConfigForDataset(dataset string) (interface{}, er
 			i = &DynamicSamplerConfig{}
 		case "EMADynamicSampler":
 			i = &EMADynamicSamplerConfig{}
+		case "RulesBasedSampler":
```
Comment (asdvalenzuela) [inline config/file_config.go:325]: The new sampler type should be added to the validation section for 'Sampler'.

File: sample/rules.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+	return nil
+}
+
+func (s *RulesBasedSampler) GetSampleRate(trace *types.Trace) (rate uint, keep bool) {
+	for _, rule := range s.Config.Rule {
+		var matched int
+
+		// no condition signifies the default
+		if rule.Condition == nil {
+			return uint(rule.SampleRate), rand.Intn(rule.SampleRate) == 0
```
Comment (asdvalenzuela) [inline sample/rules.go:30]: I think `rand.Intn` will panic if n is 0, but in the test it looked like a sample rate of 0 was valid and meant that it should be dropped?

**Review by asdvalenzuela — APPROVED** [review]

File: sample/rules.go (hunk 2 above)
Comment (martin308) [reply sample/rules.go:30]: ah good point we check for this down below but not if there is no condition. I'll add a test and fix it

File: sample/rules_test.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+			},
+			ExpectedKeep: true,
+			ExpectedRate: 10,
+		},
+		{
+			Rules: &config.RulesBasedSamplerConfig{
+				Rule: []*config.RulesBasedSamplerRule{
+					{
+						Name:       "drop",
+						SampleRate: 0,
```
[the remark this answers is not in the document]
Comment (martin308) [reply sample/rules_test.go:216]: Ah interesting! We could rethink this and have some other way of indicating that we should drop traffic?

[the remark this answers is not in the document]
Comment (martin308) [reply sample/rules_test.go:216]: I could add a drop field. I'll see what that looks like

[the remark this answers is not in the document]
Comment (martin308) [reply sample/rules_test.go]: it's in https://github.com/honeycombio/samproxy/pull/151/commits/4a6fc66d48b53e6aaf1bc995a492bd4ebbded4bd

**Pull request merged**